### PR TITLE
parse5: rework to be strongly typed

### PR DESCRIPTION
**Pull request by 43081j**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test YOUR_PACKAGE_NAME`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [x] Include [tests for your changes](https://github.com/DefinitelyTyped/DefinitelyTyped#testing)

---

This is my suggestion for #48215.

Basically the current types are really not great in all honesty. They all narrow down to `object` which makes them fairly useless for any kind of type checking as we're forced to cast to what we _think_ it is anyway. The assertion functions don't assert. Tree adapters are exported using `export =` as a hack to work around the fact you don't export an object that implements the interface.

## Changes made

* `Node` is no longer a base type - it is now a union of all the different types of node so we can correctly narrow down types with proper inference
* All the default tree adapter's types are exported from parse5 as-is (i.e. `Document` is what it is, no more hacky `| object`)
* `Attribute` is no longer a node - it isn't one, it was wrong that it ever was typed as one
* Assertion functions actually assert - `isCommentNode` actually narrows the type to `CommentNode` now
* Tree adapters are weakly typed by default (using `unknown`)
* Typed tree adapters can be defined - tree adapters which have strong typing around exposed node types

## Typed tree adapters

There's now a concept of a typed tree adapter:

```ts
interface TypeMap extends TreeAdapterTypeMap {
  node: MyNode;
  document: MyDocument;
}

declare const myAdapter: TypedTreeAdapter<TypeMap>;

// elsewhere...

const doc = parse('blah', {treeAdapter: myAdapter}); // MyDocument, NOT Document
const doc = parse('blah'); // Document
```

this is to account for the fact parse5 made the awkward decision of all tree adapters having distinct types with no cross over, and yet the same methods.

@inikulin  you wrote these types and are the sole maintainer effectively so its upto you if this rework is fine. but please do know the existing types are really unfortunate in the way they are written, the library may as well be typed as `any` right now. so even if this solution isn't used as-is, it would be good if you could help feed back how we can move towards what is "right".

@rictic i know you probably haven't been near this in eons but could be good to get your opinion if you get some time. as ill likely try use the result of this to update dom5 eventually

### CODE REVIEW

File: types/parse5/index.d.ts
Comment (ExE-Boss) [inline types/parse5/index.d.ts:659]: This should be moved into its own file.
```suggestion
```

Comment (43081j) [reply types/parse5/index.d.ts:659]: which lives where? only reason i didn't is because i wasn't sure what it'd be called since its such a deep path

Comment (ExE-Boss) [reply types/parse5/index.d.ts:659]: `types/parse5/lib/tree-adapters/default.d.ts`

**Comment by 43081j** [comment]
do note this is partially WIP too since im fully aware some of these interfaces in the tree adapter aren't as simple as inheriting the base interface. these are the problems that weren't tackled the first time around and should be now.

example:

```ts
interface Document extends parse5.Document {
```

would have trouble since a `parse5.Document` will ultimately inherit a `parse5.Node` rather than our specialised `Node`.

**Comment by 43081j** [comment]
ill carry this on at the weekend if i can. but yeah what i've now realised is:

* in fact the tree adapter's types really don't cross over with the base ones anyway, at all (in the source, these are completely different objects)
* previous types were just as wrong because tree adapter methods consume things like `Document` and `Node` while obviously no tree adapter returns those anyway (they have their own completely different versions, like htmlparser2's adapter does)

which leaves us with `parse<T>(str): T` which may as well be `parse(str) as T`. but even then, all those nice adapter methods are wrong by definition because they consume the tree adapter's document, not the base document (and not even a child interface of it...). 

so it looks like really a tree adapter's methods should consume and output the tree adapter's models, completely detached from what the default adapter's models are (parse5's own models). this sounds as weakly typed as it is and explains why we ended up with this nasty `| object` in the first place, but there will be a solution

**Comment by danger-public** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## parse5-htmlparser2-tree-adapter ([<kbd>unpkg</kbd>](https://unpkg.com/browse/parse5-htmlparser2-tree-adapter@latest/))
was missing the following properties:
1. updateNodeSourceCodeLocation
1. updateNodeSourceCodeLocation
## parse5-parser-stream ([<kbd>unpkg</kbd>](https://unpkg.com/browse/parse5-parser-stream@latest/))
was missing the following properties:
1. once
1. EventEmitter
1. once
1. EventEmitter
## parse5-serializer-stream ([<kbd>unpkg</kbd>](https://unpkg.com/browse/parse5-serializer-stream@latest/))
was missing the following properties:
1. once
1. EventEmitter
1. once
1. EventEmitter
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against d41f705b8ff4202907ec9276ec1316086847842a
</p>

**Comment by 43081j** [comment]
amazingly it now builds and tests pass.

its a breaking change but so is the version bump anyway.

i had to do some pretty strange types to achieve the "proper" types but we can blame the parse5 sources for that... to be fair it was probably written before modules were a thing.

a small summary of whats going on in the current version of this PR:

* the default tree adapter has moved to its own file
* all interfaces exported by parse5 are the "default" interfaces which the default tree adapter uses
* if you specify a tree adapter which is typed correctly using the parameterised `TreeAdapter<T>`, the right return type will be inferred for `parse`
* a tree adapter now works by specifying a map of document types it works with

for example:

```ts
interface MyTypeMap extends TreeAdapterMap {
  node: MyNode;
}

const myAdapter: TreeAdapter<MyTypeMap>;
```

this was the only cleanish way i could find to strongly type tree adapters. as they have absolutely no cross over at all, so the previous types were wrong and my first cut of this was wrong (as there's no inheritance).

**Comment by 43081j** [comment]
i have tried this out over in the polymer tooling repo. 

questions we need to answer before this can be acceptable:

* currently `childNodes` is of course `ChildNode[]` (you can't have a child node that isn't a child). however, this then means things like `childNodes.indexOf` would also require a child node, which you may not always have (you could have a `Node` only). though im tempted to side with leaving this as-is, because i _think_ the common case would be that you'd have a `ChildNode` already or something that inherits it (e.g. `Element`)

**Comment by 43081j** [comment]
just shuffled things around (again) because i noticed parent and child nodes should be a _type_, not an interface.

i.e. a child node can only be text node, element or comment node. but that is only the case for the default adapter. rather than it being a base interface those types of node inherit.

which would mean this should work:

```ts
const node: Node = whatever;

if (node.nodeName !== '#text' && node.nodeName !== '#comment') {
  // compiler now knows node is an element
}
```

and each tree adapter can specify what a "child node" is, so it could be an interface in another adapter.

**Comment by 43081j** [comment]
noticed an issue where `parse(str, { treeAdapter :T })` wasn't being inferred so i just dropped the overload magic and went with a default type parameter instead.

**Comment by 43081j** [comment]
@inikulin i know its a bit of a daunting PR but you're pretty much the only person here who can review this properly.

is there any chance you can find time to take a look? only you know the source package enough to agree or disagree that these types match up with it.

im happy to give you some examples of usage etc if that helps. but would like to avoid this going stale as these types need fixing and i have 2-3 other PRs in other repos waiting to use them

**Comment by 43081j** [comment]
IMO it is too big a change for a minor contributor to these types to sign off on FYI, we are blocked by @inikulin here.

either that or i pull someone in who uses parse5 a lot to pair with me and review it/try it out, if he isn't around/doesn't have the time.

**Comment by 43081j** [comment]
yeah this is a fun one, i think the fact its near enough a rewrite has scared everyone away.

that and the only person who can really say if its fine or not is @inikulin who hasn't been seen here yet

i've also asked for help reviewing from the polymer team as this will be used in the new dom5 release im preparing for them probably.

**Comment by rbuckton** [comment]
Part of the problem with reviewing this is that everything has been reordered, so its hard to compare the changes side-by-side. It would be easier if the new types maintained the same relative positions in the source file as their original counterparts to better detect changes.

**Comment by rbuckton** [comment]
I also think you could theoretically make `TreeAdapter` generic with a default `TreeAdapterTypeMap` that results in the same definition as `TreeAdapter` is today, possibly with something like this:.

```ts
interface TreeAdapter<T extends TreeAdapterTypeMap = typeof import("./lib/tree-adapters/default")> {
  ...
}
```

**Comment by 43081j** [comment]
I'll have a look at the ordering and give the default a go. 

Thanks for taking time to have a look. 

What do you think of the concept in general? Of the type parameter being a map of the node types. Is it a sensible solution? 

I'm pretty confident the types are fine by this point but did want a second opinion on that part

**Comment by 43081j** [comment]
@rbuckton could you take another look after christmas? if you get chance.

think it should be a better diff now

**Comment by elibarzilay** [comment]
@43081j, I think that @rbuckton's main point was that it should really have some owner review it...

**Comment by 43081j** [comment]
It definitely should. Unfortunately the sole maintainer hasn't turned up (inikulin) so I'm not too sure what else to do to make this progress. 

All other type "owners" are minor contributors so can't really review this. 

I've asked for some help in related PRs too (which depend on this fix) but no luck yet. 

Guess we can see if anyone appears after the holidays

**Comment by elibarzilay** [comment]
OK, so I'll ignore it for now -- @inikulin, any comments would help move this thing.

**Comment by andrewbranch** [comment]
@43081j I’m going to go ahead and assume that the intersection of people who are capable of giving this a real review and people who are willing to give this a real review is the empty set. I think one possible way to move forward is for you to validate and demonstrate the effect of this PR on some real-world projects. If you take a look at the dependents of @types/parse5 on npm, find some that have links back to GitHub and see if you can get a clean build with your changes here. I’d like to get an idea of how breaky this is. I don’t want all your effort to go to waste; sorry this has not been a good contributing experience. 😕

**Comment by 43081j** [comment]
Yeah it's a shame, I think the size of the change scared everyone away. 

I actually have branches of other repos waiting for this, so I will try get their maintainers to confirm all works as expected. 

I'll see if I can get a person or two to come read through the diff here too. And I'll do a build of some popular dependents (if those aren't the ones I already have branches for)

File: types/parse5/index.d.ts
Comment (LarsDenBakker) [inline types/parse5/index.d.ts:299]: Should these be `unknown`?

**Comment by 43081j** [comment]
here's some popular packages that depend on it:

- `@types/jsdom` - works fine as it only pulled in `ElementLocation` which is unchanged
- `dom5` - already have a PR waiting which uses these types at Polymer/tools#3537, works and tests pass with the new types
- `hast-util-from-parse5` - builds and passes tests, though only uses `Document`
- `gettext-extractor` - builds fine
- webhint - *fails to build* because they have their own copies of `Document` and so on, presumably to get around the previously terrible types. replacing them with the new types (or casting them) makes the build pass

while building webhint's stuff with it, i did notice `serialize` has a bug so i have fixed that now too.

i have also got some help from @LarsDenBakker to review (who may use this in the modernweb repos)

**Comment by 43081j** [comment]
i think CI is failing for unrelated reasons. a rebase will probably fix it, but ill wait until we finish tests etc so we dont mess up git history

File: types/parse5/index.d.ts
Comment (43081j) [reply types/parse5/index.d.ts:299]: i suppose this didn't exist before so its impossible anyone would be depending on the weak typing. so may be able to get away with `unknown`, ill give it a go

**Comment by 43081j** [comment]
i have no clue why jsdom fails the CI here, nothing has changed in relation to that since the last successful build of this PR.

i've tried rebasing onto master and i've double checked the diff. seems unrelated to me

maybe @rbuckton or @RyanCavanaugh can take a look (or point me at someone who can)

**Comment by LarsDenBakker** [comment]
As a large user of `parse5` and `@types/parse5` in our two monorepos: https://github.com/open-wc/open-wc and https://github.com/modernweb-dev/web in some of our most popular packages: https://www.npmjs.com/package/@web/dev-server, https://www.npmjs.com/package/@web/test-runner and https://www.npmjs.com/package/@open-wc/building-rollup I'm happy to see these improvements and we'd like to see this PR merged.

**Comment by andrewbranch** [comment]
Thanks @LarsDenBakker! To clarify, have you tried out these changes on those projects already? If so, did you have any type errors that needed to be fixed? How many? Were the errors usefully pointing out unsoundness, or were they just noise? Were they easy to fix or did they take a lot of effort?

**Comment by LarsDenBakker** [comment]
@andrewbranch yes I've tried it out. Because of the current state of the parse5 types, I had a lot of type casts and ts-ignores. Removing those gives me type errors, which I can now fix correctly because the new types actually make sense. Especially type return values and assertions work nicely now.

**Review by andrewbranch — APPROVED** [review]
Thanks so much for vetting this on a number of existing dependents @43081j and @LarsDenBakker. That seems as good or better than an owner review to me. It also helps that this coincidentally corresponds with a major version bump, so users won’t get any breaks until they upgrade to a new major version anyway, which seems reasonable. And most of all, thanks for your patience, @43081j!

**Comment by 43081j** [comment]
Thanks a lot for helping me out. And you too @LarsDenBakker.

I have PRs ready for a couple of dependent libraries once this merges too. I'll keep a close eye on it in case anyone opens issues for it in future. 

Let's see if typescript bot can handle this not being it's own comment:

Ready to merge

**Comment by 43081j** [comment]
Ready to merge 👀

**Pull request merged**
